File: Caldera/src/model/C3Pot_FineTuning.js
import { POT_MAX, POT_MIDPOINT } from "./DigiPotModel.js";

export function fineTuning(controller) {
  const sensorDeadband = 100;
  const sensorLow = controller.constructor.SENSOR_MIDPOINT - sensorDeadband / 2;
  const sensorHigh = controller.constructor.SENSOR_MIDPOINT + sensorDeadband / 2;
  const midStep = 150;
  const wiperLow = Math.trunc((POT_MIDPOINT - midStep / 2) / 2);
  const wiperHigh = POT_MAX - wiperLow;

  if (controller.zone !== controller.Zone.inZone) {
    controller.inZone = false;
    controller.phase = controller.Phase.SEARCH;
    return;
  }

  let direction = 0;
  const wiperLevel = controller.mid.getLevel();

  if (wiperLevel < wiperLow) direction = 1;
  else if (wiperLevel > wiperHigh) direction = -1;

  if (direction !== 0) {
    controller.top.offsetLevel(direction);
    controller.bot.offsetLevel(direction);
    controller.mid.offsetLevel(direction * midStep);
    return;
  }

  const sensorValue = controller.lastSensorValue();

  if (sensorValue < sensorLow) controller.mid.offsetLevel(1);
  else if (sensorValue > sensorHigh) controller.mid.offsetLevel(-1);
  else controller.inZone = true;
}

File: Caldera/other/1. TIA/C3Pot_FineTuning.cpp
#include "C3Pot.h"
#include "CUSB.h"

void C3Pot::fineTuning() {
  static constexpr int SENSOR_DEADBAND = 100;
  static constexpr int SENSOR_LOW      = CAutoPot::SENSOR_MIDPOINT - SENSOR_DEADBAND/2;
  static constexpr int SENSOR_HIGH     = CAutoPot::SENSOR_MIDPOINT + SENSOR_DEADBAND/2;

  static constexpr int MID_STEP = 150;

  static constexpr int WIPER_LOW  = (CAutoPot::POT_MIDPOINT - MID_STEP/2) / 2;
  static constexpr int WIPER_HIGH = 255 - WIPER_LOW;


  if (zone != Zone::inZone) { inZone = false; phase = Phase::SEARCH; return; }

  int direction = 0;
  int wiperLevel = mid.getLevel();

  if (wiperLevel < WIPER_LOW ) direction = +1;
  else
  if (wiperLevel > WIPER_HIGH) direction = -1;

  if (direction != 0) {
    top.offsetLevel(direction);
    bot.offsetLevel(direction);
    mid.offsetLevel(direction * MID_STEP);
    return;
  }

  
  int sensorValue = lastSensorValue();

  if (sensorValue < SENSOR_LOW ) mid.offsetLevel(+1);
  else  
  if (sensorValue > SENSOR_HIGH) mid.offsetLevel(-1);
  else
    inZone = true;
  
}